core/matrix/coo.cpp

```diff
@@ -187,8 +187,13 @@ void Coo<ValueType, IndexType>::read(const mat_data& data)
 template <typename ValueType, typename IndexType>
 void Coo<ValueType, IndexType>::read(const device_mat_data& data)
 {
-    // make a copy, read the data in
-    this->read(device_mat_data{this->get_executor(), data});
+    this->set_size(data.get_size());
+    this->values_ = make_const_array_view(
+        data.get_executor(), data.get_num_elems(), data.get_const_values());
+    this->col_idxs_ = make_const_array_view(
+        data.get_executor(), data.get_num_elems(), data.get_const_col_idxs());
+    this->row_idxs_ = make_const_array_view(
+        data.get_executor(), data.get_num_elems(), data.get_const_row_idxs());
```

Comment:
I think just simple copies should also work  ?
```suggestion
    this->values_ = data.arrays.values;
    this->col_idxs_ = data.arrays.col_idxs;
    this->row_idxs_ = data.arrays.row_idxs;
```

Reply:
`arrays` isn't a member of data, and the internal arrays are not accessible to `Csr`, so this needs to stay. Though it is a bit confusing, since assigning a non-const view to an array causes it to be reassigned, while this causes a copy